### PR TITLE
WIP: ENH: Polygon Surface Area Calculation

### PR DESCRIPTION
**Pull request by tylerjereddy**
Just continuing from the previous [WIP PR](https://github.com/scipy/scipy/pull/7515) from a clean branch (various git/ github issues with the other...). Still very much a debug stage / work in progress PR.

At the moment, `test_subtriangles_simple()` is still failing, but the point that splits the test spherical triangle into three equal area subtriangles looks quite reasonable:

spherical_triangle_split_example

The question now is why the test is failing to generate the correct areas, and why each of the subtriangles has a very different surface area at the moment (not expected based on the diagram above).

It could just be that the new subtriangles have a property that stresses the code in a way that previous simpler unit tests did not. Not sure.

### CODE REVIEW

**Comment by james-nichols** [comment]
Commenting here to get following. Currently working with Tyler to check our testing code and help further develop the routine

**Comment by tylerjereddy** [comment]
@james-nichols Awesome! Just to keep the discussion open my latest ( wild ? ) hypothesis is that all of my previous test cases include a "polar" (N pole) vertex & this may cause an issue because the NASA JPL whitepaper specifically excludes polygons that include a pole (I'm not clear on whether a vertex being a pole counts as a pole being "inside" the polygon). 

This may partly explain the factor of two discrepancy in the spherical polygon area implementation vs. their paper, but probably still some things I'm not quite getting right...

**Comment by tylerjereddy** [comment]
I haven't committed any changes but I did refactor the code locally to explicitly print out the integrated areas under each edge of a given spherical polygon. Taking one of the subtriangles from `test_subtriangles_simple` does seem to reveal an issue. Consider the following debug output from one of the subtriangles on a unit sphere that does not include a pole (simplifies the issues):

```
subtriangle_vertices: [[-1.          0.          0.        ]
 [ 0.          1.          0.        ]
 [-0.57735027  0.57735027  0.57735027]]
lambda_vals: [ 3.14159265  1.57079633  2.35619449]
phi_vals: [ 0.          0.          0.61547971]
area between vertices:  [-1.  0.  0.] [-0.57735027  0.57735027  0.57735027]
and south pole is: 1.01212308393
area between vertices:  [ 0.  1.  0.] [-1.  0.  0.]
and south pole is: -1.57079632679
area between vertices:  [-0.57735027  0.57735027  0.57735027] [ 0.  1.  0.]
and south pole is: 1.01212308393
actual_subtriangle_area: 0.453449841059
expected_sub_area: 0.5235987755982988
```

The reported areas under the edges to the "South Pole" of a unit sphere are used in the derivation itself, so they are sensible targets for debugging. While the `-pi / 2` "integration" area matches expectations, the other two areas (which involve the trilateration point, D) do not match the intuitive area expectation.

The subtriangle vertices printed above represent a triangle that should have an area of pi / 6. We'd then expect that the edges involving the trilateration point enclose a Southern area that is 0.5 * pi / 6 + pi / 4, with the latter pi / 4 representing the area contribution below the equator. This works out to 1.04719.... instead of 1.0121... and using the former value produces the subtriangle area expected by the unit test.

The question is: is this a trigonometric precision issue or something more sinister? Looks a bit too large to be precision-related, but maybe not.

The temporarily refactored / debug code looks like this:

```python
 43 def _spherical_polygon_area(vertices, radius):
 44     num_vertices = vertices.shape[0]                                                               
 45     area_sum = 0
 46     lambda_vals = np.arctan2(vertices[...,1], vertices[...,0]) # longitudes                        
 47     phi_vals = np.arcsin(vertices[...,2] / float(radius), dtype=np.float64) # latitudes            
 48     print("lambda_vals:", lambda_vals)
 49     print("phi_vals:", phi_vals)    
 50         
 51     for i in xrange(num_vertices):
 52         #forward_index, backward_index = _vertex_index_strider(i, num_vertices)                    
 53         delta_lambda = (lambda_vals[i] -
 54                         lambda_vals[i-1])
 55         second_term = 2 + np.sin(phi_vals[i]) + np.sin(phi_vals[i-1])                              
 56         area_element = delta_lambda * second_term * (radius ** 2) * 0.5                            
 57         print("area between vertices: ", vertices[i], vertices[i-1])                               
 58         print("and south pole is:", area_element)             
 59         area_sum += area_element
 62     area = area_sum                        
 63     return area   
```

**Comment by tylerjereddy** [comment]
I've confirmed that the Python [spherical_geometry library](https://pypi.python.org/pypi/spherical-geometry) gives the exact area of the subtriangle expected & I don't think there are any fundamental issues with the way the code is written here (both the trilateration / triangle splitting & area calculation code look ok).

Rather, the problems seem to be that:
1. this approach is an approximation
2. it doesn't handle cases where the vertices include a N/S pole.

Since the particular case above doesn't include a pole, we are only left with the issue of this being an approximation. More specifically, the area under an edge (great circle arc) of the spherical polygon is approximated using the average value of phi (N-S) between two vertices. However, to get a more accurate result we may need to discretize the integration into many very small pieces. 

Although for longitude (E-W) the discretization might be possible via a trivial `np.linspace` for the latitude (N-S) we need to travel along the great circle arc between two vertices--so we need to interpolate along great circle arcs at the appropriate North-South elevation.

Assuming that the above interpolation can be accomplished in a straightforward way, one might envision a user-specified parameter for the number of iterations / area integral discretizations based on their desired balance between performance and accuracy. The relationship between accuracy and the number of discretizations may also vary with the size of the spherical polygon (lengths of the arcs / edges, etc.).

We should probably try the above before thinking about handling polar-inclusive cases by i.e., translating the polygon on the surface of the sphere first, etc.

**Comment by tylerjereddy** [comment]
The latest commit adds "discretized" integration to the spherical polygon area calculation -- allowing user-specified accuracy of the area calculation at the cost of iterations (smaller integration elements under the great circle arcs).

The results are pretty much exactly as one would expect with more iterations (pieces) providing more accurate areas & most unit tests are now passing, including the months-long effort to get the subtriangle division test to work.

There should now be quite a bit more low-hanging fruit for improvement / code clean-up / test expansions after overcoming the source of confusion (the approximation / discretization issue).

**Comment by tylerjereddy** [comment]
- following test migration to  `pytest` format and removing dependence on `hypothesis` testing library

```7 failed, 55 passed, 5 warnings in 8.22 seconds```

- some of that is probably related to North / South pole vertex tests and / or the Cython code not yet having been adjusted for area discretization

**Comment by tylerjereddy** [comment]
- Expanded the unit tests a bit & improved support for spherical polygons not on the unit sphere (r != 1)
- temporarily disabled the switch to Cython spherical polygon SA calc code to simplify things for the moment (in this early stage of development)

Some things I'd like to get working / sorted out soon:
- [ ] the full spectrum of parametrized tests in `test_subtriangles_iterative()` -- any amount of iterations > 1 currently fails (could be a silly oversight in the test itself)
- [ ] determine why `TestSimpleAreas.test_half_hemisphere_area_reverse_order` currently has issues & if we can remedy them with current approach

**Comment by tylerjereddy** [comment]
The first issue is resolved now -- I was just being too demanding on the sizes of the subtriangles. An expected area of (1 / (3 ^ 10)) of a quater hemisphere is tolerated, but (1 / (3 ^ 55)) of a quarter hemisphere is clearly a very small spherical triangle so I've scaled that back to the more reasonable limit and it works now for that parametrized test.

**Comment by tylerjereddy** [comment]
- all unit tests pass locally after the latest commit / fixes
- here are a few suggestions for next logical steps:
  - [ ] unify to a single (probably Cython / C) implementation of the core iterative code moving forward so that two parallel implementations (Python / Cython) don't have to be developed side-by-side all the time
  - [ ] add a unit test for a spherical polygon that includes the South Pole (0 0 -1) as a vertex -- I suspect this may cause problems & require further handling improvements in the code
  - [ ] add unit tests for spherical polygons that *contain* the N/S pole(s) (not just as vertices, but internally) -- this should stress the code & require more adaptations
  - [ ]  add a "functional" unit test that uses `SphericalVoronoi` to generate a bunch of pseudo-random spherical polygons & make sure the sum of all calculated polygon areas is equivalent to the theoretical surface area of the original sphere
  - [ ] maybe clean-up / rebase a bit so that we can occasionally run the CI test suite, though CI skipping is probably still mostly advised in the near future because of WIP
- there are other things to think about too ("internal" vs. "external" areas of bounded spherical polygons), but one step at a time

File: scipy/spatial/_polygon_surface_area.py [hunk 1]
```diff
@@ -0,0 +1,153 @@
...
+    forward_index = index + 1
+    backward_index = index - 1
+    if forward_index >= num_vertices:
+        forward_index = 0
+    return forward_index, backward_index
+
+def _planar_polygon_area(vertices):
+    num_vertices = vertices.shape[0]
+    area_sum = 0
+    for i in xrange(num_vertices):
```
Comment (xoviat) [inline scipy/spatial/_polygon_surface_area.py]: Have you profiled this? How long does this code take? It might be worth moving to Cython.

Comment (tylerjereddy) [reply scipy/spatial/_polygon_surface_area.py]: Yeah, see the checklist above -- I think I'll just switch to Cython / C-only moving forward. I tried to develop in Python / Cython side-by-side (to compare the speed of the two) but the dual implementation approach during development eventually got neglected and I mostly used Python for convenience.

That said, we're probably still quite far from "production ready" to worry about performance too much.

**Comment by tylerjereddy** [comment]
One somewhat-pathological scenario is the infinite set of great circles that divide a sphere into two equal-area hemispheres. From a design standpoint, how should we handle this? 

For the case of the equatorial great circle, it is ambiguous if the North or South hemisphere is specified, though this is largely irrelevant from the standpoint of the absolute value of the surface area returned.

Theoretically, any set of two antipodes could define such a great circle. In my testing so far, it is easier to handle the equatorial split than the East / West split, but we also need to think of all possible great circles & detect / handle those appropriately. And should there be a minimum number of vertices required (3?) ?

**Comment by tylerjereddy** [comment]
@james-nichols Thoughts on the issues here? Am I trying to extend too far beyond the original whitepaper, which does make some concessions with respect to spherical polygons that encompass a pole?

**Comment by tylerjereddy** [comment]
The most recent commit drafts an attempt to handle spherical polygons that are great circles (split the sphere into hemispheres) by checking for vertex membership in a plane that contains the origin.

While this seems like a nice idea (return 1/2 sphere surface area if great circle is detected), my unit tests are already detecting a problem scenario.

Consider vertices at North, East and West of a sphere: [0,0,1], [1,0,0], [-1,0,0]

I had originally pictured this as **half of a hemisphere**, but it is equally valid to connect the East / West vertices via a great circle arc that goes through the South Pole vs. the Equator, meaning that the area may also be interpreted as that of an entire hemisphere.

My question / challenge is -- can I robustly handle these ambiguities?

**Comment by tylerjereddy** [comment]
One possible idea: if any consecutive vertices are connected by a geodesic of length `pi`, reject the spherical polygon as ambiguous (infinite possible great circles connect those two vertices). In other words, reject consecutive antipodes.

**Comment by tylerjereddy** [comment]
All unit tests pass locally after incorporating consecutive antipode rejection + hemisphere (plane) checking. I think that's good (or at least, time to think of more edge cases!).

**Comment by tylerjereddy** [comment]
I rebased against the latest version of master so that CI can occasionally run, though will still mostly endeavor to [ci skip] at this stage. I do note that in some parts of the test matrix the full testsuite can pass, including the tests from this WIP branch, but there's still a long way to go (and mess to clean up).

The next thing I might like to do is add a "functional test" -- the result polygons from SphericalVoronoi as input and the area of the original input sphere system as the expected "sum of areas" of all polygons. That will, by definition, include polygons that contain the poles (a known weakness of the approach used here) so could be helpful in finding residual problems.

It would be nice to start the effort to "vectorize" the code so that it operates on an array of input polygons instead of the current single-polygon input approach, but I'm still hesitant to make that jump until I'm feeling really confident that the single-polygon testing scenario / handling is rock-solid.

**Comment by tylerjereddy** [comment]
I drafted the functional test based on `SphericalVoronoi`. Currently it fails to reconstitute the surface area of the entire unit sphere from its constitutent Voronoi polygons based on random generators.

It will be interesting to determine why this fails.

**Comment by tylerjereddy** [comment]
Here's a visualization of a "problematic" spherical polygon from the functional test using `SphericalVoronoi` that generates random spherical polygons. Problematic is effectively defined as having an unusually large result for surface area -- in this case a quarter of the surface area of the entire sphere, which it clearly should not be:

debug_sv_sa_failure

In this case, notice that two out of the five vertices are effectively indistinguishable (0 and 4 are right on top of each other).

Indeed, this can be seen in the raw coord data:
```python
 16 suspicious_polygon = np.array([[-0.79278828,-0.48489777,-0.36927076],
 17                                [-0.95264505,-0.01427026, 0.3037495 ],
 18                                [-0.92458779, 0.35065228, 0.14893085],
 19                                [-0.91491027, 0.40075481,-0.0483196 ],
 20                                [-0.79277493,-0.48322543,-0.37148498]])
```

I'm guessing this isn't a coincidence. One complicating matter is that some of the other suspicious polygons don't seem to share this property in a comprehensive sense:

```python
suspicious polygon_area: 2.4009947972
suspicious polygon: [[-0.46697717  0.40065984  0.78829183]
 [-0.65596415  0.58294079  0.47946956]
 [-0.92458779  0.35065228  0.14893085]
 [-0.95264505 -0.01427026  0.3037495 ]
 [-0.74729965 -0.05701915  0.66203629]
 [-0.65194264  0.0126658   0.7581625 ]]
suspicious polygon_area: 1.46301679085
suspicious polygon: [[-0.81452972  0.36315966 -0.45238965]
 [-0.77539244 -0.43201758 -0.46057287]
 [-0.77381038 -0.4380721  -0.45750446]
 [-0.79277493 -0.48322543 -0.37148498]
 [-0.91491027  0.40075481 -0.0483196 ]
 [-0.85660871  0.49523509 -0.14478858]
 [-0.8326736   0.49176277 -0.25460568]]
suspicious polygon_area: 1.21243475735
suspicious polygon: [[-0.81452972  0.36315966 -0.45238965]
 [-0.77539244 -0.43201758 -0.46057287]
 [-0.67684209  0.0433663  -0.73484974]
 [-0.62798157  0.37767849 -0.68043964]
 [-0.75109864  0.39409346 -0.52966138]]
suspicious polygon_area: 1.20409873488
suspicious polygon: [[-0.47865337 -0.40994121 -0.77642717]
 [-0.77381038 -0.4380721  -0.45750446]
 [-0.77539244 -0.43201758 -0.46057287]
 [-0.67684209  0.0433663  -0.73484974]
 [-0.25883378  0.32266286 -0.91043602]
 [-0.15934674  0.19509311 -0.96775374]]
```

In any case, the test seems to be doing what it should -- probing potential weaknesses in the algorithm. Now the question is whether they can be addressed in a straightforward way.

**Comment by tylerjereddy** [comment]
It seems there was an issue with lambda values striding + / - the value of `pi` such that `2 * pi` had to be added or subtracted back in prior to the area integration to avoid integrating a massive area when two points were on either side of the equatorial boundary.

After applying the above patch, I find one more suspicious category of polygon:

debug_sv_sa_failure

The large point is the North Pole. If you ignore the fact that matplotlib doesn't handle spherical polygons very well, it certainly looks like this problem case falls into one of the major categories described as problematic by NASA: "containing a pole."

So, if we want to gracefully handle this scenario it looks like we'll have to pre-check all input polygons for containing one or both poles. So we add the point in polygon problem, with a spherical twist, on top of the surface area calculation!

**Comment by tylerjereddy** [comment]
[This discussion](https://blog.element84.com/determining-if-a-spherical-polygon-contains-a-pole.html) looks relevant to the problem, albeit potentially a fair bit of work to implement properly without slowing things down too much.

**Comment by tylerjereddy** [comment]
I've added the first draft of a new function (and associated unit test) to detect whether a spherical polygon contains the North or South Pole. The test doesn't pass yet & indeed this also won't currently handle a few corner cases including a spherical polygon that would contain *both* poles, but one thing at a time.. will try to get North Pole containment detection working first.

**Comment by tylerjereddy** [comment]
`pole_in_polygon()` now [apparently] correctly detects spherical polygons that contain the North Pole as an interior point (unit test passes).

Some low hanging fruit for improvement:
- enhance `test_pole_in_polygon()` with spherical polygon input test cases that:
  - contain the South pole
  - contain neither North nor South pole
  - the above input should be straightforward to generate using `SphericalVoronoi` with random generators + visual inspection

Once the above tests pass, the next logical task is likely to incorporate `pole_in_polygon()` into the spherical polygon area calculation workflow. The question is whether we should attempt to "rotate" the problematic pole-containing polygon until it no longer contains a pole or if we should return i.e., NaN for these polygons which are not tractable by the general algorithm.

**Comment by tylerjereddy** [comment]
- test_pole_in_polygon() has been expanded to include the extra appropriate test cases described above & each case passes

**Comment by tylerjereddy** [comment]
We can now handle spherical polygons that contain the North or South Pole (!), as long as they can be successfully rotated away from that pole on the surface of the sphere. All 106 unit tests pass locally, including the complete reconstitution of the surface area of a sphere based on its constituent spherical polygons generated randomly by `SphericalVoronoi`.

Steps from here:
- start to think about vectorization / speed up to handle large numbers of planar / spherical polygons simultaneously (finally !)
- start to clean up the code (single Cython / C implementation instead of the mostly pure Python stuff used here for prototyping with half-baked Cython stuff on the side)
  - related: speed up unit tests: 331 seconds in serial on an i9 machine
- one of the unit tests stochastically has floating point issues -- make that a little less stringent
- documentation (related to the eventual API / cleaned up code)

**Comment by tylerjereddy** [comment]
`pole_in_polygon` migrated to Cython (crudely) -- tests continue to pass; no substantial perf improvement with that migration though lots of room for refinement.

**Comment by tylerjereddy** [comment]
`_spherical_polygon_area()` migrated to Cython with 13-fold reduction in total test suite time for `test_polygon_SA.py`

**Comment by tylerjereddy** [comment]
The dual Cython / Python implementations / flags are now removed -- there's just a single Cython implementation of the code to perform area calculations etc. This was a long-standing issue, so that's good to have resolved & also prevents the number of units tests run from appearing too inflated because of the dual implementations.

If someone is interested in chipping in, it should be possible to add code / test to detect & handle spherical polygons that contain both the North and South poles by [using an approach similar to that for single pole detection](https://blog.element84.com/determining-if-a-spherical-polygon-contains-a-pole.html) -- I think the follow-up link to that article has some clues, although the follow-up link is currently broken in my hands.

**Comment by tylerjereddy** [comment]
I've added crude support for vectorization (return an array of areas given multiple input polygons). A functional test passes for this.

I note that supporting vectorization of this nature means trying to perform rapid calculations on a ragged matrix (polygons w/ different numbers of vertices), such that normal ufuncs aren't really useful.

The current iteration over polygons happens in Python -- should at least be able to make some gains by moving that over to Cython soon.

**Comment by tylerjereddy** [comment]
A [working link](https://blog.element84.com/follow-up-to-determining-if-a-spherical-polygon-contains-a-pole.html) for the discussion of double pole containment within polygons is now available too -- so I or someone should be able to catch the double-pole case now too.

**Comment by tylerjereddy** [comment]
500 spherical polygon surface area calculations currently requires ~0.04 seconds

**Comment by tylerjereddy** [comment]
I've added a test case with a spherical polygon that contains both poles--see below (forgiving matplotlib for not knowing this polygon is on a sphere):

image

This was actually a tricky case to conceive. We can't simply use a polygon that encompasses > a left or right hempishere, because the algorithm assumes that we have the smallest contained area. A thin rectangular strip that contains both poles struck me as the easiest to generate using [slerp](https://en.wikipedia.org/wiki/Slerp) (spherical linear interpolation).

Currently, `pole_in_polygon()` does not correctly discern the case of dual pole containment -- it should likely be modified to detect it as [previously described](https://blog.element84.com/follow-up-to-determining-if-a-spherical-polygon-contains-a-pole.html).

**Comment by tylerjereddy** [comment]
If I use the `_slerp` code I wrote for area calculation to prepare the spherical polygon for plotting, you can see that things look much nicer:

image

Perhaps I'll briefly mention spherical linear interpolation as an alternative to the spherical polygon plotting solutions using triangulations being considered in matplotlib.

**Comment by tylerjereddy** [comment]
~0.036 seconds for 500 spherical polygons now; I should perhaps be cautious of time invested in performance though -- at least until other parts of the code are more mature.

**Comment by tylerjereddy** [comment]
The approach used in the NASA JPL paper to assess whether a spherical polygon contains a pole requires prior knowledge of a point outside the spherical polygon, so does not strike me as an appropriate approach to assessing dual-pole containment.

At this stage I'm tempted to accept the inability to detect a spherical polygon that contains both poles & to move forward with refinements. One could argue that a common use case would be several spherical polygons on the same sphere, such that any vertex of another spherical polygon should be a legit candidate for an "external" point, but there's really nothing restricting input from containing multiple spherical polygons from many different spheres with matching radii, etc.

I've not had much success implementing the follow-up blog post approach cited above, although the original (more detailed) blog post was just fine for detecting single pole containment in a spherical polygon.

**Comment by tylerjereddy** [comment]
Other useful "point in spherical polygon" references

- [Bevis and Chatelain, 1989](https://link.springer.com/article/10.1007/BF00894449) -- requires prior knowledge of a point inside the polygon; uses classic ray crossing
- [Li et al., 2017](https://link.springer.com/chapter/10.1007/978-3-319-69487-0_5) -- apparently tries to avoid the conventionally expensive projection to the plane before applying ray crossing

**Comment by tylerjereddy** [comment]
I read through the more recent manuscript above -- looks suitable for all spherical point in polygon situations & claims to be ~5 orders of magnitude faster than the previous work in empirical tests.

The implementation looks a little tricky & includes a variable resolution grid data structure; perhaps most appropriate to split that off into another PR first.

**Comment by tylerjereddy** [comment]
I started some early work on the Li et al. (2017) algorithm here: https://github.com/tylerjereddy/grid-point-sphere

As usual, going from the manuscript description to a fully functional & tested algorithm is likely to take some work, but if that is successful it would probably make the current challenges in this PR more tractable. The plan would probably be to first open a PR for "point in spherical polygon" at some sufficient level of maturity, and as usual it would be subject to discussion on the merit for inclusion, etc.

**Comment by tylerjereddy** [comment]
The initial code to cast the first level of the grid has been drafted in the above repo, along with initial attempt to register which grid cells contain edges of the spherical polygon.

Casting the deeper level grids depends on subdividing only grid cells with edges inside them -- this may be a use case for sparse data structures that I'm less familiar with; I'll likely proceed by somehow organizing the different level grids in a dictionary or something of that nature for now.

**Comment by tylerjereddy** [comment]
Closing this PR of mine -- almost 2000 lines of new code & the primary discussion is just with myself in comments above over last two years.

I may try to break this out into smaller PRs as useful utilities to build up comp. geometry capabilities, but depends on demonstrating sufficient global usefulness for inclusion in SciPy.

**Comment by rgommers** [comment]
@tylerjereddy I won't be much help with WIP discussions as I don't know much about computational geometry, but if some of the smaller pieces are ready for review, I'll try to help!

**Comment by tylerjereddy** [comment]
I've been making some progress on the point in spherical polygon data structure described in 2017, with my library able to produce this example plot now:

level_1_grid

I haven't succeeded in getting funding for it (yet), so it will continue to move slowly, most likely. 

The area problem is partly addressed by some of Peter's work I need to review (in the convex case, where you already have a Voronoi diagram on the sphere), but the point-in-polygon algorithm seems useful as well for i.e., airspace monitoring etc., given authors' claim of 5 orders of magnitude speedup over 1989 algorithm.

**Pull request closed without merging**